src/hooks/useConfigurePatch.ts

```diff
@@ -0,0 +1,246 @@
+import { queryString, string } from "utils";
+import { convertArrayToObject, mapStringArrayToObject } from "utils/array";
+
+const { omitTypename } = string;
+const { parseQueryString } = queryString;
+
+type ConfigurePatchState = {
+  description: string;
+  selectedBuildVariants: string[];
+  selectedBuildVariantTasks: VariantTasksState;
+  patchParams: ParameterInput[];
+  selectedTab: number;
+  disableBuildVariantSelect: boolean;
+  selectedDownstreamPatches: DownstreamPatchState;
+};
+
+type Action =
+  | { type: "setDescription"; description: string }
+  | { type: "setSelectedBuildVariants"; buildVariants: string[] }
+  | { type: "setPatchParams"; params: ParameterInput[] }
+  | { type: "setSelectedBuildVariantTasks"; variantTasks: VariantTasksState }
+  | { type: "setSelectedTab"; tabIndex: number }
+  | {
+      type: "updatePatchData";
+      description: string;
+      buildVariants: string[];
+      params: ParameterInput[];
+      variantTasks: VariantTasksState;
+      downstreamPatches: DownstreamPatchState;
+    }
+  | {
+      type: "setSelectedDownstreamPatches";
+      downstreamPatches: DownstreamPatchState;
+    };
+
+const initialState = ({ selectedTab = 0 }: { selectedTab: number }) => ({
+  description: "",
+  selectedBuildVariants: [],
+  selectedBuildVariantTasks: {},
+  patchParams: null,
+  selectedTab,
+  disableBuildVariantSelect: tabToIndexMap[selectedTab] === PatchTab.Tasks,
+  selectedDownstreamPatches: {},
+});
+
+const reducer = (state: ConfigurePatchState, action: Action) => {
+  switch (action.type) {
+    case "setDescription":
+      return {
+        ...state,
+        description: action.description,
+      };
+    case "setSelectedBuildVariants":
+      return {
+        ...state,
+        selectedBuildVariants: action.buildVariants,
+      };
+    case "setSelectedBuildVariantTasks":
+      return {
+        ...state,
+        selectedBuildVariantTasks: action.variantTasks,
+      };
+    case "setSelectedDownstreamPatches":
+      return {
+        ...state,
+        selectedDownstreamPatches: action.downstreamPatches,
+      };
+    case "setPatchParams":
+      return {
+        ...state,
+        patchParams: omitTypename(action.params),
+      };
+    case "setSelectedTab": {
+      let tab = indexToTabMap.indexOf(PatchTab.Tasks);
+      if (action.tabIndex !== -1 && action.tabIndex < indexToTabMap.length) {
+        tab = action.tabIndex;
+      }
+      return {
+        ...state,
+        selectedTab: tab,
+        disableBuildVariantSelect:
+          indexToTabMap[action.tabIndex] !== PatchTab.Tasks,
+      };
+    }
+    case "updatePatchData":
+      return {
+        ...state,
+        description: action.description,
+        selectedBuildVariants: action.buildVariants,
+        patchParams: omitTypename(action.params),
+        selectedBuildVariantTasks: action.variantTasks,
+        selectedDownstreamPatches: action.downstreamPatches,
+      };
+
+    default:
+      throw new Error();
```

Comment:
Can we throw a more detailed error here? So its easier to trace down errors 
`Unrecognized reducer action ${action.type}`

Reply:
bump.